pkg/pillar/cmd/domainmgr/domainmgr.go

```diff
@@ -1330,42 +1334,95 @@ func doActivateTail(ctx *domainContext, status *types.DomainStatus,
 // throws back error stating that the device is busy if enabling the vlan
 // filtering is tried immediately after bridge creation.
 // We can either loop retrying in zedrouter or enable it when needed in domainmgr
-func enableVlanFiltering(bridgeName string) {
+func enableVlanFiltering(bridgeName string) error {
 	bridge, err := netlink.LinkByName(bridgeName)
 	if err != nil {
 		log.Errorf("enableVlanFiltering: LinkByName failed for %s: %s", bridgeName, err)
-		return
+		return err
 	}
 	if !*bridge.(*netlink.Bridge).VlanFiltering {
 		// Enable VLAN filtering on bridge
 		if err := netlink.BridgeSetVlanFiltering(bridge, true); err != nil {
-			errStr := fmt.Sprintf("enableVlanFiltering on %s failed: %s",
+			err = fmt.Errorf("enableVlanFiltering on %s failed: %s",
 				bridgeName, err)
-			log.Errorf(errStr)
+			log.Error(err)
+			return err
 		}
 	}
+	return nil
 }
 
-func setupVlans(vifList []types.VifInfo) {
+func setupVlans(vifList []types.VifInfo) error {
+	deadline := time.Now().Add(10 * time.Second)
+	const delay = 500 * time.Millisecond
 	for _, vif := range vifList {
+		if vif.Vlan.End == 0 {
+			// VLAN not configured for this interface
+			continue
+		}
+
+		var err error
 		bridgeName := vif.Bridge
-		enableVlanFiltering(bridgeName)
+		for {
+			err := enableVlanFiltering(bridgeName)
+			if err == nil {
+				break
+			}
+			if errors.Is(err, unix.EBUSY) {
+				if time.Now().Before(deadline) {
+					log.Warnf("setupVlans: bridge %s is busy, will retry in %s",
+						bridgeName, delay)
+					time.Sleep(delay)
+					continue
+				}
+			}
+			return err
+		}
 
-		link, err := netlink.LinkByName(vif.Vif)
-		if err != nil {
-			log.Errorf("setupVlans: Vlan setup failed: %s", err)
-			return
+		var link netlink.Link
+		for {
+			link, err = netlink.LinkByName(vif.Vif)
+			if err == nil {
+				if link.Attrs().MasterIndex == 0 && time.Now().Before(deadline) {
+					log.Warnf("setupVlans: interface %s is not yet bridged, will retry in %s",
+						vif.Vif, delay)
+					time.Sleep(delay)
+					continue
+				}
+				break
```

Comment:
Should this break be considered a failure if MasterIndex == 0?

Reply:
In testing I noticed that with Xen there is often a small delay between interface appearing and getting bridged, hence we only need to wait a bit.

Reply:
good point, will add there an explicit error, otherwise it will fail with ENOTSUPP which may not be that clear.

Reply:
done